File: ICPC/Training 10-9/C.py
def calculate_pisano_period(p):
    if p == 2:
        return 3
    elif p == 5:
        return 20

    if p % 5 == 1 or p % 5 == 4:
        period = p - 1
    elif p % 5 == 2 or p % 5 == 3:
        period = 2 * p + 2

    return period

def gcd(a, b):
    while b:
        a, b = b, a % b
    return a

def calculate_pisano_period_large(p):
    if p <= 1:
        return 0
    
    period = 1
    a, b = 1, 1

    while not (a == 1 and b == 0):
        a, b = b, (a + b) % p
        period += 1

    return period

p = int(input())

if p <= 5:
    result = calculate_pisano_period(p)
else:
    if p % 5 == 0:
        result = calculate_pisano_period_large(p)
    else:
        result = calculate_pisano_period(p)

print(result)

File: ICPC/Training 10-9/C.cpp
#include <iostream>
#include <unordered_map>

using namespace std;
using ll = long long;

unordered_map<ll, ll> pisano_periods;

ll calculate_pisano_period_large(ll p)
{
    ll a = 0, b = 1, c = a + b;
    for (ll i = 0; i < p * p; i++)
    {
        c = (a + b) % p;
        a = b;
        b = c;
        if (a == 0 && b == 1)
            return i + 1;
    }
    return 0;
}

void precompute_pisano_periods()
{
    for (ll p = 2; p <= 1000000; p++)
    {
        pisano_periods[p] = calculate_pisano_period_large(p);
    }
}

int main()
{
    precompute_pisano_periods();

    ll p;
    cin >> p;

    if (p <= 5)
    {
        if (p == 2)
            cout << "3\n";
        else if (p == 5)
            cout << "20\n";
        else
            cout << "0\n";
    }
    else
    {
        ll result = pisano_periods[p];
        cout << result << "\n";
    }

    return 0;
}
